File: traps.py
"""有一块 MxN 的土地, 雨后积水, 有水标记为'w', 干燥为'.'.八联通的积水被认为是连在一起的, 求院子里有多少水洼."""
"""为了计算院子里有多少个水洼，我们可以使用深度优先搜索（DFS）或广度优先搜索（BFS）来遍历整个土地"""
"""这里使用DFS深搜"""


class Traps:
    def __init__(self):
        self.grid = []
        self.M, self.N = None, None
        self.traps = 0

    def get_grid(self):
        """获取长宽及土地列表"""
        self.M, self.N = map(lambda x: int(x), input('输入 长&高: ').split(' '))
        self.grid = [list(input(f'第{i + 1}行,共{self.N}行: ')) for i in range(self.N)]

    def find_w(self):
        """遍历土地 若找到水洼 则进行dfs"""
        for i in range(self.N):
            for j in range(self.M):
                if self.grid[i][j] == 'w':
                    # 如果该地为水洼, 则执行dfs
                    self.dfs(i, j)
                    # 搜索完水洼后将水洼数+1
                    self.traps += 1

    def dfs(self, x, y):
        """深搜 dfs"""
        # 搜索后更改, 防止重复搜索
        self.grid[x][y] = '.'
        # 搜索八个方向
        for i in range(-1, 2):
            for j in range(-1, 2):
                # 防止数组越界
                if 0 <= x + i < self.N and 0 <= y + j < self.M:
                    # 若发现水洼, 再次进行dfs
                    if self.grid[x + i][y + j] == 'w':
                        self.dfs(x+i, y+j)

    def run(self):
        self.get_grid()
        self.find_w()
        print(f"Number of water traps: {self.traps}")


if __name__ == '__main__':
    Traps().run()

File: traps.cpp
#include <iostream>  
#include <vector>  
  
using namespace std;  
  
// 定义八个方向的偏移量  
int dx[] = {-1, -1, -1, 0, 0, 1, 1, 1};  
int dy[] = {-1, 0, 1, -1, 1, -1, 0, 1};  
  
// 函数原型声明  
void dfs(vector<vector<char>>& grid, int x, int y, int M, int N);  
  
int numWaterTraps(vector<vector<char>>& grid) {  
    if (grid.empty() || grid[0].empty()) return 0;  
  
    int M = grid.size();  
    int N = grid[0].size();  
    int waterTraps = 0;  
  
    // 遍历整个土地  
    for (int i = 0; i < M; ++i) {  
        for (int j = 0; j < N; ++j) {  
            // 如果当前位置是水洼，则进行深度优先搜索  
            if (grid[i][j] == 'w') {  
                dfs(grid, i, j, M, N);  
                ++waterTraps; // 发现一个新的水洼  
            }  
        }  
    }  
  
    return waterTraps;  
}  
  
// 深度优先搜索函数  
void dfs(vector<vector<char>>& grid, int x, int y, int M, int N) {  
    // 检查边界和是否访问过  
    if (x < 0 || x >= M || y < 0 || y >= N || grid[x][y] != 'w') {  
        return;  
    }  
  
    // 标记当前位置为已访问  
    grid[x][y] = '.';  
  
    // 对八个方向进行递归搜索  
    for (int i = 0; i < 8; ++i) {  
        int nx = x + dx[i];  
        int ny = y + dy[i];  
        dfs(grid, nx, ny, M, N);  
    }  
}  
  
int main() {  
    int num = 0;
    int width, height;
    cin >> width;
    cin >> height;

    vector<vector<char>>grid;
    vector<char>line;
    for(int i = 0; i < height; i++){
        for(int j = 0; j <= width; j++){
            line.push_back(cin.get());
        }
        grid.push_back(line);
        line.clear();
    }
    
    // 计算水洼数量  
    int numWater = numWaterTraps(grid);  
  
    // 输出结果  
    cout << "Number of water traps: " << numWater << endl;  
  
    return 0;  
}
